### PR TITLE
Feature/#8_Foe_and_BasicFoe_classes

### PR DESCRIPTION
**Pull request by allantsai3**

### CODE REVIEW

File: src/game_object/enemies.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
+import pygame
+from pygame.locals import *
+
+from src.const import *
+from src.game_object.dynamic import *
+
+class enemies(Dynamic):
+    depth = -4
```
Comment (ozcer) [inline src/game_object/enemies.py]: I changed how depth is declared and drawn. See other classes for example

File: src/game_object/enemies.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+        self.alive = True
+
+    def draw(self):
+        super().draw()
+
+    def update(self):
+        self.x += self.dx
+        self.y += self.dy
+        self.rect.center = self.x, self.y
```
Comment (ozcer) [inline src/game_object/enemies.py]: This part is taken care of by the parent class Dyanmic. Call super().update()

File: src/game_object/enemies.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
+import pygame
+from pygame.locals import *
+
+from src.const import *
+from src.game_object.dynamic import *
+
+class enemies(Dynamic):
+    depth = -4
+    def __init__(self, game):
```
Comment (ozcer) [inline src/game_object/enemies.py]: We want to at least have the option to set its position

File: src/game_object/enemies.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        self.game = game
+
+        # sprite
+        self.sprites = {"ooze": pygame.image.load("sprites/monster1.png")}
+        self.type = {"basic"}
+
+        self.image = self.sprites["ooze"]
+
+        self.x = DISPLAY_WIDTH + 50
+        self.y = DISPLAY_HEIGHT/2
```
Comment (ozcer) [inline src/game_object/enemies.py]: not hard code it to spawn off screen

File: src/game_object/enemies.py (hunk 1 above)
Comment (ozcer) [inline src/game_object/enemies.py]: I changed how depth is declared and drawn. See other classes for example

File: src/game_object/enemies.py (hunk 2 above)
Comment (ozcer) [inline src/game_object/enemies.py]: This part is taken care of by the parent class Dynamic. Call super().update()

File: src/game_object/enemies.py (hunk 3 above)
Comment (allantsai3) [reply src/game_object/enemies.py]: done

File: src/game_object/enemies.py (hunk 4 above)
Comment (allantsai3) [reply src/game_object/enemies.py]: done

File: src/game_object/enemies.py (hunk 1 above)
Comment (allantsai3) [reply src/game_object/enemies.py]: done

File: src/game_object/Foe/basicfoe.py [hunk 5]
```diff
@@ -1,20 +1,26 @@
...
         self.image = random.choice(list(self.sprites.items()))[1]
 
         #Movement
-        self.dx = -1
+        self.dx = -5
         self.dy = 0
 
         #Monster Hp
-        self.life = 50
-        super().__init__(game, self.image)
```
Comment (ozcer) [inline src/game_object/Foe/basicfoe.py]: Leave the params empty. I believe they are passed implicitly.

File: src/game_object/Foe/basicfoe.py [hunk 6]
```diff
@@ -1,20 +1,26 @@
...
+        self.hp = 50
+
+        super().__init__(game, pos, depth, self.image)
```
Comment (ozcer) [inline src/game_object/Foe/basicfoe.py]: same here

File: src/game_object/Foe/basicfoe.py [hunk 7]
```diff
@@ -1,20 +1,26 @@
...
 class BasicFoe(Foe):
-    def __init__(self, game):
+    def __init__(self,
+                 game,
+                 pos=(DISPLAY_WIDTH + 50, DISPLAY_HEIGHT/2),
+                 depth=-4):
         #Sprite
         self.sprites = {"ooze": pygame.image.load("sprites/foesprites/foe1.png"),
                         "tedders": pygame.image.load("sprites/foesprites/foe2.png")}
         self.image = random.choice(list(self.sprites.items()))[1]
```
Comment (ozcer) [inline src/game_object/Foe/basicfoe.py]: I feel like there's a more Pythonic way to do this. But I just can't find one :/

File: src/game_object/Foe/foe.py [hunk 8]
```diff
@@ -1,18 +1,21 @@
-import pygame
```
Comment (ozcer) [inline src/game_object/Foe/foe.py]: lower case foe for directory name

File: src/game_object/Foe/foe.py [hunk 9]
```diff
@@ -26,6 +29,15 @@ def draw(self):
...
-        self.rect.center = self.x, self.y
+        super().update()
+
+        #health condition
+        if self.hp is 0 or self.decayable():
+            self.alive = False
+
+        if self.alive is False:
+            self.kill()
+            print("monster deleted")
```
Comment (ozcer) [inline src/game_object/Foe/foe.py]: remove debug prints in production(master branch) we don't want people to see this in our "product" :)

File: src/game_object/Foe/foe.py [hunk 10]
```diff
@@ -26,6 +29,15 @@ def draw(self):
...
-        self.x += self.dx
-        self.y += self.dy
-        self.rect.center = self.x, self.y
+        super().update()
+
+        #health condition
+        if self.hp is 0 or self.decayable():
+            self.alive = False
+
+        if self.alive is False:
```
Comment (ozcer) [inline src/game_object/Foe/foe.py]: the Pythonic way is `if not self.alive`

File: src/game_object/game_object.py [hunk 11]
```diff
@@ -28,4 +30,13 @@ def show_attributes(self):
...
+
+
+    def decayable(self):
+        """
+        check if is too far out from main surface
+        :return: bool
+        """
+        # active zone = main surface x 2
+        active_zone = self.game.surface.get_rect().inflate(self.rect.width*2, DISPLAY_HEIGHT/2)
+        return not active_zone.colliderect(self.rect)
```
Comment (ozcer) [inline src/game_object/game_object.py]: newline end of file

File: main.py [hunk 12]
```diff
@@ -1,121 +1,124 @@
...
         self.entities[class_name].add(object)
-        
+
         # also add to global sprite group
         self.entities[GLOBAL_SPRITE_GROUP].add(object)
-    
+
     def make_walls(self):
-        gap_height = random.randint(GAP_SIZE/2, DISPLAY_HEIGHT-GAP_SIZE/2)
+        gap_height = random.randint(GAP_SIZE / 2, DISPLAY_HEIGHT - GAP_SIZE / 2)
```
[the remark this answers is not in the document]
Comment (allantsai3) [reply main.py]: autoformat op

File: main.py [hunk 13]
```diff
@@ -1,121 +1,124 @@
...
-        bottom_wall = Wall(self, (DISPLAY_WIDTH+WALL_WIDTH, gap_bottom + bottom_height / 2), (WALL_WIDTH, bottom_height))
+        bottom_wall = Wall(self, (DISPLAY_WIDTH + WALL_WIDTH, gap_bottom + bottom_height / 2),
+                           (WALL_WIDTH, bottom_height))
         self.add_entity(bottom_wall)
 
-        
-        
+
 if __name__ == "__main__":
     Game()
```
[the remark this answers is not in the document]
Comment (allantsai3) [reply main.py]: autoformat not op

File: main.py [hunk 14]
```diff
@@ -1,121 +1,124 @@
...
 import os
-import random
+import sys
+
 import pygame
 from pygame.locals import *
 
+from src.HUD import HUD
 from src.const import *
+from src.game_object.Foe.basicfoe import *
```
[the remark this answers is not in the document]
Comment (allantsai3) [reply main.py]: yes sir

File: src/game_object/Foe/foe.py (hunk 9 above)
Comment (allantsai3) [reply src/game_object/Foe/foe.py]: yes sir

File: src/game_object/Foe/foe.py (hunk 10 above)
Comment (allantsai3) [reply src/game_object/Foe/foe.py]: okie

File: src/game_object/game_object.py (hunk 11 above)
Comment (allantsai3) [reply src/game_object/game_object.py]: k

File: src/game_object/foe/basic_foe.py [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+        self.dx = -5
+        self.dy = 0
+
+        #Monster Hp
+        self.hp = 50
+
+        super().__init__(game, pos, depth, self.image)
+
+    def draw(self):
+        Foe.draw(self)
```
Comment (ozcer) [inline src/game_object/foe/basic_foe.py]: super()

File: src/game_object/foe/basic_foe.py [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+
+    def update(self):
+        Foe.update(self)
```
Comment (ozcer) [inline src/game_object/foe/basic_foe.py]: super()

File: src/game_object/foe/foe.py [hunk 17]
```diff
@@ -0,0 +1,42 @@
+from src.game_object.dynamic import Dynamic
+
+class Foe(Dynamic):
+    def __init__(self,
+                 game,
+                 pos,
+                 depth,
```
Comment (ozcer) [inline src/game_object/foe/foe.py]: everything except game and pos should have defaults

**Comment by ozcer** [comment]
make sure you squash your commits into one commit with concise summary, then merge

**Pull request merged**